### ISSUE TITLE
Triage: Possible memory leak

### ISSUE THREAD

**Issue by jheck90** [kieraneglin/pinchflat#130]
First off, really loving the application.

I left pinchflat to do its thing over night, as it all seems pretty good to go running on Nomad on my homelab. In checkout out the metrics this morning, I noticed the memory kept creeping up, then maxed out, then shot back down. Any idea what's going on there?

<img width="940" alt="image" src="https://github.com/kieraneglin/pinchflat/assets/35469139/4ca78e7a-a9db-46f1-9e3f-f36f076cf0bf">

This is alot of what my logs are full of:
```bash

22:01:47.192 [error] Exqlite.Connection (#PID<0.1977.0>) disconnected: ** (DBConnection.ConnectionError) client #PID<0.2017.0> timed out because it queued and checked out the connection for longer than 15000ms

#PID<0.2017.0> was at location:

    (exqlite 0.19.0) lib/exqlite/sqlite3.ex:102: Exqlite.Sqlite3.multi_step/3
    (exqlite 0.19.0) lib/exqlite/sqlite3.ex:140: Exqlite.Sqlite3.try_fetch_all/3
    (exqlite 0.19.0) lib/exqlite/sqlite3.ex:134: Exqlite.Sqlite3.fetch_all/3
    (exqlite 0.19.0) lib/exqlite/connection.ex:672: Exqlite.Connection.get_rows/2
    (exqlite 0.19.0) lib/exqlite/connection.ex:618: Exqlite.Connection.execute/4
    (db_connection 2.6.0) lib/db_connection/holder.ex:354: DBConnection.Holder.holder_apply/4
    (db_connection 2.6.0) lib/db_connection.ex:1512: DBConnection.run_execute/5

22:35:32.832 [info] {"args":{"id":12},"id":370,"meta":{},"system_time":1711406132831790727,"max_attempts":20,"queue":"media_collection_indexing","worker":"Pinchflat.SlowIndexing.MediaCollectionIndexingWorker","source":"oban","event":"job:start","attempt":12,"tags":["media_source","media_collection_indexing"]}
22:35:32.859 [info] [yt-dlp] called with: https://www.youtube.com/@SesameStreet --simulate --skip-download --ignore-no-formats-error --print-to-file %(.{id,title,was_live,webpage_url,description,aspect_ratio,duration,upload_date})j /tmp/pinchflat/data/c86265ac80577f6e5ad68b6cc10462abe2b6b9181f10f218dbed3e0ab2d004f4.json
22:37:55.131 [error] Exqlite.Connection (#PID<0.1977.0>) disconnected: ** (DBConnection.ConnectionError) client #PID<0.2017.0> timed out because it queued and checked out the connection for longer than 15000ms

```

**Comment by kieraneglin** [comment]
That's really interesting! I would expect this pattern somewhat as indexing + download jobs kickoff then are completed as `yt-dlp` does use a fair bit of RAM when it's working - especially since the result of the index is (temporarily) held in memory as it's being processed.

That said, that's about 2-3x the RAM usage I would expect to see. I'm _assuming_ the SQLite error is also due to the RAM exhaustion issue since it tracks to me that the DB couldn't be opened if there isn't enough free memory to do so.

This next part is mostly me thinking out loud for future context + debugging ideas. The Erlang VM is very good at keeping memory in check, but the big caveat there is that guarantees aren't maintained when working with a long-running NIF _or port_. It could also be that I need to hibernate the file-follower genserver (or add a `fullsweep_after`)

TODO:

- Look into enabling live dashboard in "prod" to let people see memory usage per-process
- Look into culling zombie processes (see `Port` module docs)
- Look into streaming the indexing file one complete instead of reading directly into memory
- Look into hibernating the file follower genserver ([context](https://elixirforum.com/t/extremely-high-memory-usage-in-genservers/4035/23))

---

I don't have a satisfactory answer at the moment, but I'm going to look into this. There's a chance that it is from `yt-dlp` acting as "normal" (or a memory leak from `yt-dlp`), but even then I'd like to see if I can tweak some knobs to curb that behaviour.

Thanks for the report!

**Comment by jheck90** [comment]
@kieraneglin Let me know if you need any other data/logs!

**Comment by kieraneglin** [comment]
I was able to look into this some more and I was able to reproduce it fairly easily. It looks like this is from `yt-dlp` and it is considered "normal" for it to consume upwards of 2GB of memory for a sufficiently large channel. I wasn't seeing that in my testing, but then I realized I was only indexing channels with a max of ~800 videos

I'm not sure if there's a way I can mitigate this, but I'm going to keep looking into it. In the meantime I recommend allocating a little more ram to the container (although this is more of a bandaid than a fix). I've also enabled a new `/dev/dashboard` endpoint that can provide some troubleshooting information if RAM usage keeps climbing.

**Comment by somethingGoneWrong** [comment]
Hi, I've been using the program for a few days now and I've noticed intensive CPU and RAM usage, the downward spike at 4:00 AM is due to the automatic update of the docker image.

![screen](https://github.com/kieraneglin/pinchflat/assets/40141099/c1446727-9224-483b-bcd9-9a5a80e2d50e)


maybe I'm putting too many youtube channels to control? Should I put less?

![screen#4](https://github.com/kieraneglin/pinchflat/assets/40141099/8b2949de-d65d-4d15-8722-ef68da28190a)


I see that it goes continuous prompts as if it continuously updates them all, while in the settings I selected daily update.
I'll leave a few more pictures, if needed I can post logs.
Pinchflat has the purple color line:

![screen#2](https://github.com/kieraneglin/pinchflat/assets/40141099/3fde706f-9299-4d6a-93a5-d9c404b8e802)

![screen#3](https://github.com/kieraneglin/pinchflat/assets/40141099/cb493ab5-6acc-4fee-bd82-72ddede0d3a9)

![screen#1](https://github.com/kieraneglin/pinchflat/assets/40141099/fc45082f-ad7f-4c36-bb1d-cfe1286d6579)

**Comment by kieraneglin** [comment]
@somethingGoneWrong I've been looking into it and this is almost certainly from the fact that TEDx has >200k videos. I didn't even know channels that large existed 😅. This is going to present two main issues.

Issue the first: 

Since indexing alone would take well over a day (regardless of whether you're downloading or not), here's what's probably happening:

- indexing runs all day
- nightly restart
- indexing has to restart from the beginning because there isn't a checkpointing system
- repeat forever

So it'll never stop indexing the same content. I'll look into some sort of checkpointing system to minimize reindexing of known content on app restart. Given how yt-dlp works I'm not sure if this will actually solve the issue, but it's worth checking out!

Second issue:

yt-dlp has to keep some level of internal metadata in RAM while it's scanning as well as storing the metadata the app uses for indexing. The latter is negligible (< 1kB per video), but I've seen estimates that the former can take _up to_ 120kB per video which would result in around 24GB of RAM in the worst case! Again, a similar solution to the first point could work here, depending on how yt -dlp stores its metadata. For a bit of context, yt-dlp is basically a glorified screen scraper so it behaves a little unintuitively when it comes to fetching only a subset of videos. 

Anyway, I'll see what I can find out!

**Comment by somethingGoneWrong** [comment]
@kieraneglin I don't know if it is due to an update or the fact that I deleted the tedx channel but now the ram consumption has become normal, it is around 500-600 MB, while the CPU consumption for me remains too high:

![screen52](https://github.com/kieraneglin/pinchflat/assets/40141099/d7185a09-6a0a-44b8-9783-033e252f29aa)

![screen54](https://github.com/kieraneglin/pinchflat/assets/40141099/ff4f3c18-0834-4c21-b713-9e644a33b490)

Pinchflat is the purple line, if it is helpful I found a couple of such errors:

```
05:35:15.946 [info] [yt-dlp] called with: https://www.youtube.com/watch?v=xiSBGTQtOME --no-simulate --no-progress --windows-filenames --format-sort res:720,+codec:avc:m4a --output /downloads/link4universe_%(upload_date>%Y-%m-%d)S_%(title)S.%(ext)S --print-to-file after_move:%()j /tmp/pinchflat/data/bd206dbf89f9bcfcbc352c8cdb577eb07ffb5b3964ccd7534d06679e928e8ee0.json
05:35:48.502 [info] {"args":{"id":372832},"id":1183,"meta":{},"state":"success","max_attempts":20,"queue":"media_fetching","worker":"Pinchflat.Downloading.MediaDownloadWorker","source":"oban","duration":32561491,"event":"job:stop","queue_time":939398,"attempt":1,"tags":["media_item","media_fetching"]}
06:07:09.895 [error] GenServer {Oban.Registry, {Oban, Oban.Stager}} terminating
** (Exqlite.Error) Database busy
UPDATE "oban_jobs" AS o0 SET "state" = ? WHERE (o0."id" IN (?))
    (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1054: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:952: Ecto.Adapters.SQL.execute/6
    (oban 2.17.3) lib/oban/engines/lite.ex:132: Oban.Engines.Lite.stage_jobs/3
    (oban 2.17.3) lib/oban/engine.ex:229: anonymous fn/4 in Oban.Engine.stage_jobs/3
    (oban 2.17.3) lib/oban/engine.ex:315: anonymous fn/3 in Oban.Engine.with_span/4
    (telemetry 1.2.1) /app/deps/telemetry/src/telemetry.erl:321: :telemetry.span/3
    (oban 2.17.3) lib/oban/stager.ex:132: anonymous fn/2 in Oban.Stager.check_leadership_and_stage/1
    (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1358: anonymous fn/3 in Ecto.Adapters.SQL.checkout_or_transaction/4
Last message: :stage
06:07:10.923 [info] {"args":{},"id":1180,"meta":{},"system_time":1711858030922826476,"max_attempts":20,"queue":"local_metadata","worker":"Pinchflat.Boot.DataBackfillWorker","source":"oban","event":"job:start","attempt":1,"tags":["media_item","media_metadata","local_metadata","data_backfill"]}
06:07:10.923 [info] Running data backfill worker
```
I think using sqlite for this workload is not enough, it would be to try using pg as an alternative.
thanks for the app.

**Comment by kieraneglin** [comment]
I haven't pushed any updates so the RAM usage was certainly from TEDx - I'm still looking into ways to mitigate that usage and I'll let you know if I find anything!

I suspect that cpu usage is expected - ultimately yt-dlp hands off to ffmpeg to assemble the final video and ffmpeg _loves_ cpu. If that's the culprit, that's actually a sign things are well tuned to use those resources. It should drop down pretty substantially once it's done processing all the videos in its queue. 

re: SQLite, I definitely think that's my fault rather than a shortfall of SQLite. PF really isn't that database intensive and SQLite is surprisingly capable when tuned well. I will look into adjusting things, but is there any chance you're storing the config folder on a network share?

**Comment by somethingGoneWrong** [comment]
> Edit: re-reading that error, I actually don't think that one's a big deal. I didn't realize the app just booted. That one's on my radar and I think I know the fix, but it should be recoverable. Still curious about the network share thing!

No, the config is in the local disk, the download folder is in a NFS storage.
```
/share_docker/pf_config/db:
total 205M   
 262163 drwxr-xr-x    2 dockerus dockerus    4.0K Mar 30 04:06 .
 262162 drwxr-xr-x    6 dockerus dockerus    4.0K Mar 26 12:29 ..
 262164 -rw-r--r--    1 dockerus dockerus  198.5M Apr  1 13:19 pinchflat.db
 262168 -rw-r--r--    1 dockerus dockerus   32.0K Apr  1 13:20 pinchflat.db-shm
 262167 -rw-r--r--    1 dockerus dockerus    6.1M Apr  1 13:20 pinchflat.db-wal

/share_docker/pf_config/extras:
total 8K     
 262892 drwxr-xr-x    2 dockerus dockerus    4.0K Mar 26 12:29 .
 262162 drwxr-xr-x    6 dockerus dockerus    4.0K Mar 26 12:29 ..
 262893 -rw-r--r--    1 dockerus dockerus       0 Mar 30 04:06 .keep
 262894 -rw-r--r--    1 dockerus dockerus       0 Mar 26 12:29 cookies.txt

/share_docker/pf_config/logs:
total 3M     
 262165 drwxr-xr-x    2 dockerus dockerus    4.0K Mar 23 18:51 .
 262162 drwxr-xr-x    6 dockerus dockerus    4.0K Mar 26 12:29 ..
 262166 -rw-r--r--    1 dockerus dockerus    2.6M Apr  1 13:18 pinchflat.log

/share_docker/pf_config/metadata:
total 24K    
 262169 drwxr-xr-x    4 dockerus dockerus    4.0K Mar 24 10:58 .
 262162 drwxr-xr-x    6 dockerus dockerus    4.0K Mar 26 12:29 ..
 262170 -rw-r--r--    1 dockerus dockerus       0 Mar 30 04:06 .keep
 262174 drwxr-xr-x  359 dockerus dockerus   12.0K Apr  1 10:56 media_items
 262171 drwxr-xr-x   55 dockerus dockerus    4.0K Apr  1 10:56 sources
```

**Comment by kieraneglin** [comment]
I think I've found the source of a possible memory leak and it's from a zombie process created when a job runner is terminated prematurely. This wouldn't happen in all cases but is more likely to happen when working with large channels. I'm looking into a fix for that 👍 

I haven't had any success in finding an approach that would make it more efficient or allow checkpointing when indexing a massive channel like TEDx. I'll pick that back up to try again one of these weekends.

**Comment by kieraneglin** [comment]
I have a fix for the zombie process issue in #182 that'll be part of the next release later today or tomorrow.

With that, I'm going to close this issue because I believe that's all I can do to ensure memory is freed. If it turns out this doesn't solve things, please let me know and I'll re-open it if that ends up being the case (and something within my control)!

I'll create a separate issue for tracking the checkpointing of large channels once I pick that issue back up.

Thanks once again for the report!
